File: src/main/java/lambda/evaluator/QuoteFunctionEvaluator.java
package lambda.evaluator;

import lambda.core.datamodels.Node;
import lambda.core.datamodels.ProgramStackItem;
import lambda.core.datamodels.Stack;

interface QuoteFunctionEvaluator {

    void evaluateQuoteFunction(
        ProgramStackItem top,
        Stack<Node> evalStack,
        Stack<ProgramStackItem> programStackItem
    );

    static QuoteFunctionEvaluator newInstance(PostEvaluationStackUpdater postEvaluationStackUpdater) {
        return (top, evalStack, programStack) -> {
            final var quoteExprNode = top.getFunctionExpressionNode();
            final var quoted = quoteExprNode.getChildren().get(1);
            postEvaluationStackUpdater.updateStacksAfterEvaluation(
                quoted,
                top.getVariableMap(),
                evalStack,
                programStack
            );
        };
    }
}

File: src/main/java/lambda/evaluator/QuoteFunctionEvaluator.kt
package lambda.evaluator

import lambda.core.datamodels.Stack
import lambda.core.datamodels.Node
import lambda.core.datamodels.ProgramStackItem

internal interface QuoteFunctionEvaluator {
    fun evaluateQuoteFunction(
        top: ProgramStackItem,
        evalStack: Stack<Node>,
        programStack: Stack<ProgramStackItem>
    )
}

internal class QuoteFunctionEvaluatorImpl(
    private val postEvaluationStackUpdater: PostEvaluationStackUpdater
): QuoteFunctionEvaluator {

    override fun evaluateQuoteFunction(
        top: ProgramStackItem,
        evalStack: Stack<Node>,
        programStack: Stack<ProgramStackItem>
    ) {
        val quoteExprNode = top.functionExpressionNode
        val quoted = quoteExprNode.children[1]
        return postEvaluationStackUpdater.updateStacksAfterEvaluation(
            quoted,
            top.variableMap,
            evalStack,
            programStack
        )
    }
}